File: Sources/SomnioServerCore/Connections/PeriodicLoop.swift
import Foundation
import Logging
import ServiceLifecycle

/// Shared periodic-loop skeleton for the server's tick services. Sleeps for `interval` between
/// passes and runs `body` each wake. On graceful shutdown the inner sleep cancels and the loop
/// returns normally (the `CancellationError` branch); any other sleep failure is logged under
/// `"\(label) sleep failed"` and also returns. Callers that need work after the loop ends (e.g.
/// a final save) put it after the `await runPeriodically(...)` call.
func runPeriodically(
    interval: Duration,
    logger: Logger,
    label: String,
    _ body: @escaping @Sendable () async -> Void
) async {
    await cancelWhenGracefulShutdown {
        while !Task.isCancelled {
            do {
                try await Task.sleep(for: interval)
            } catch is CancellationError {
                return
            } catch {
                logger.warning(
                    "\(label) sleep failed",
                    metadata: ["error": "\(error)"]
                )
                return
            }
            await body()
        }
    }
}

File: packages/server/src/services/periodicLoop.ts
/** Resolves after `ms`, or immediately (with `false`) once `signal` aborts. */
export function sleep(ms: number, signal: AbortSignal): Promise<boolean> {
  return new Promise((resolve) => {
    if (signal.aborted) {
      resolve(false);
      return;
    }
    const timer = setTimeout(() => {
      signal.removeEventListener('abort', onAbort);
      resolve(true);
    }, ms);
    const onAbort = () => {
      clearTimeout(timer);
      resolve(false);
    };
    signal.addEventListener('abort', onAbort, { once: true });
  });
}

/**
 * The shared periodic-loop skeleton for the tick services: sleeps for `intervalMs` between
 * passes and runs `body` each wake; an abort during the sleep ends the loop cleanly. Work that
 * must follow the last pass (a final save) goes after the returned promise.
 */
export async function runPeriodically(intervalMs: number, signal: AbortSignal, body: () => Promise<void>): Promise<void> {
  while (await sleep(intervalMs, signal)) {
    await body();
  }
}
